File: src/test/java/im/tox/tox4j/NetworkTest.java
package im.tox.tox4j;

import im.tox.tox4j.enums.ToxProxyType;
import im.tox.tox4j.enums.ToxStatus;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;

import static org.junit.Assert.*;

public class NetworkTest extends ToxCoreImplTestBase {

    private static final int TOX_COUNT = 10;

    private void testBootstrap(String ipv4, int port, byte[] dhtId) throws Exception {
        try (ToxCore tox = newTox()) {
            long start = System.currentTimeMillis();
            tox.bootstrap(ipv4, port, dhtId);
            ConnectedListener status = new ConnectedListener();
            tox.callbackConnectionStatus(status);
            while (!status.isConnected()) {
                tox.iteration();
                try {
                    Thread.sleep(tox.iterationInterval());
                } catch (InterruptedException e) {
                    // Probably the timeout was reached, so we ought to be killed soon.
                }
            }
            long end = System.currentTimeMillis();
            if (LOGGING) System.out.println("Bootstrap to remote bootstrap node took " + (end - start) + "ms");
        }
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrap4() throws Exception {
        assumeIPv4();
        testBootstrap(nodes[0].ipv4, nodes[0].port, nodes[0].dhtId);
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrap6() throws Exception {
        assumeIPv6();
        testBootstrap(nodes[0].ipv6, nodes[0].port, nodes[0].dhtId);
    }

    @Test(timeout = TIMEOUT)
    public void testBootstrapSelf() throws Exception {
        // TODO: don't know how to test this on localhost
    }

    @Test(timeout = TIMEOUT)
    public void testLANDiscoveryAll() throws Exception {
        try (ToxList toxes = new ToxList(TOX_COUNT)) {
            long start = System.currentTimeMillis();
            // TODO: Generous timeout required for this; should be made more reliable.
            while (!toxes.isAllConnected()) {
                toxes.iteration();
                try {
                    Thread.sleep(toxes.iterationInterval());
                } catch (InterruptedException e) {
                    // Probably the timeout was reached, so we ought to be killed soon.
                }
            }
            long end = System.currentTimeMillis();
            if (LOGGING) System.out.println("Connecting all of " + toxes.size() + " toxes with LAN discovery " +
                    "took " + (end - start) + "ms");
        }
    }

    @Test(timeout = TIMEOUT)
    public void testLANDiscoveryAny() throws Exception {
        try (ToxList toxes = new ToxList(TOX_COUNT)) {
            long start = System.currentTimeMillis();
            while (!toxes.isAnyConnected()) {
                toxes.iteration();
                try {
                    Thread.sleep(toxes.iterationInterval());
                } catch (InterruptedException e) {
                    // Probably the timeout was reached, so we ought to be killed soon.
                }
            }
            long end = System.currentTimeMillis();
            if (LOGGING) System.out.println("Connecting one of " + toxes.size() + " toxes with LAN discovery " +
                    "took " + (end - start) + "ms");
        }
    }

}

File: projects/tox4j/src/test/java/im/tox/tox4j/core/NetworkTest.scala
package im.tox.tox4j.core

import im.tox.tox4j.DhtNodeSelector.node
import im.tox.tox4j.TestConstants.Timeout
import im.tox.tox4j._
import im.tox.tox4j.core.NetworkTest.logger
import im.tox.tox4j.core.ToxCoreFactory.{withTox, withToxes}
import org.scalatest.FlatSpec
import org.scalatest.concurrent.Timeouts
import org.scalatest.time.SpanSugar._
import org.slf4j.{Logger, LoggerFactory}

import scala.language.postfixOps

object NetworkTest {
  private val logger = LoggerFactory.getLogger(classOf[NetworkTest])
  private val ToxCount = 10
}

final class NetworkTest extends FlatSpec with Timeouts {

  private def testBootstrap(ipv6Enabled: Boolean, udpEnabled: Boolean, ip: String, port: Int, dhtId: Array[Byte]): Unit = {
    val action = s"bootstrap to remote node on $ip:$port with ${if (udpEnabled) "UDP" else "TCP"}${if (ipv6Enabled) 6 else 4}"

    withTox(ipv6Enabled, udpEnabled) { tox =>
      logger.info(s"Attempting to $action")
      val start = System.currentTimeMillis

      if (!udpEnabled) {
        tox.addTcpRelay(ip, port, dhtId)
      }
      tox.bootstrap(ip, port, dhtId)

      val status = new ConnectedListener
      tox.callback(status)
      while (!status.isConnected) {
        tox.iterate(())
        Thread.sleep(tox.iterationInterval)
      }

      val end = System.currentTimeMillis

      logger.info(s"Success: $action took ${end - start} ms")
    }
  }

  "bootstrap" should "connect with UDP4" in {
    assume(ToxCoreTestBase.checkIPv4.isEmpty)
    failAfter(Timeout millis) {
      testBootstrap(ipv6Enabled = false, udpEnabled = true, node.ipv4, node.udpPort, node.dhtId)
    }
  }

  it should "connect with UDP6" in {
    assume(ToxCoreTestBase.checkIPv6.isEmpty)
    failAfter(Timeout millis) {
      testBootstrap(ipv6Enabled = true, udpEnabled = true, node.ipv6, node.udpPort, node.dhtId)
    }
  }

  it should "connect with TCP4" in {
    assume(ToxCoreTestBase.checkIPv4.isEmpty)
    failAfter(Timeout millis) {
      testBootstrap(ipv6Enabled = false, udpEnabled = false, node.ipv4, node.tcpPort, node.dhtId)
    }
  }

  it should "connect with TCP6" in {
    assume(ToxCoreTestBase.checkIPv6.isEmpty)
    failAfter(Timeout millis) {
      testBootstrap(ipv6Enabled = true, udpEnabled = false, node.ipv6, node.tcpPort, node.dhtId)
    }
  }

  "LAN discovery" should "connect all nodes" in {
    failAfter(Timeout millis) {
      withToxes(NetworkTest.ToxCount) { toxes =>
        val action = s"Connecting all of ${toxes.size} toxes with LAN discovery"
        logger.info(action)

        val start = System.currentTimeMillis

        while (!toxes.isAllConnected) {
          toxes.iterate(())
          Thread.sleep(toxes.iterationInterval)
        }

        val end = System.currentTimeMillis

        logger.info(s"$action took ${end - start} ms")
      }
    }
  }

  it should "connect at least one instance" in {
    failAfter(Timeout millis) {
      withToxes(NetworkTest.ToxCount) { toxes =>
        val action = s"Connecting one of ${toxes.size} toxes with LAN discovery"
        logger.info(action)

        val start = System.currentTimeMillis

        while (!toxes.isAnyConnected) {
          toxes.iterate(())
          try {
            Thread.sleep(toxes.iterationInterval)
          } catch {
            case e: InterruptedException =>
          }
        }

        val end = System.currentTimeMillis

        logger.info(s"$action took ${end - start} ms")
      }
    }
  }

}
